qiskit/circuit/qpy_serialization.py

```diff
@@ -1454,38 +1577,88 @@ def _write_circuit(file_obj, circuit):
+    processed_qubit_indices = set()
+    processed_clbit_indices = set()
+
+    def process_qregs(buf, qregs, in_circuit=True):
+        for reg in qregs:
             standalone = all(bit._register is reg for bit in reg)
             reg_name = reg.name.encode("utf8")
-            file_obj.write(struct.pack(REGISTER_PACK, b"q", standalone, reg.size, len(reg_name)))
-            file_obj.write(reg_name)
-            REGISTER_ARRAY_PACK = "!%sI" % reg.size
-            file_obj.write(struct.pack(REGISTER_ARRAY_PACK, *(qubit_indices[bit] for bit in reg)))
-        for reg in circuit.cregs:
+            buf.write(
+                struct.pack(REGISTER_V4_PACK, b"q", standalone, reg.size, len(reg_name), in_circuit)
+            )
+            buf.write(reg_name)
+            REGISTER_ARRAY_PACK = "!%sq" % reg.size
+            bit_indices = []
+            for bit in reg:
+                bit_index = qubit_indices.get(bit, -1)
+                if bit_index in processed_qubit_indices:
+                    bit_index = -1
+                if bit_index >= 0:
+                    processed_qubit_indices.add(bit_index)
+                bit_indices.append(bit_index)
+            buf.write(struct.pack(REGISTER_ARRAY_PACK, *bit_indices))
+
+    def process_cregs(buf, cregs, in_circuit=True):
+        for reg in cregs:
             standalone = all(bit._register is reg for bit in reg)
             reg_name = reg.name.encode("utf8")
-            file_obj.write(struct.pack(REGISTER_PACK, b"c", standalone, reg.size, len(reg_name)))
-            file_obj.write(reg_name)
-            REGISTER_ARRAY_PACK = "!%sI" % reg.size
-            file_obj.write(struct.pack(REGISTER_ARRAY_PACK, *(clbit_indices[bit] for bit in reg)))
+            buf.write(
+                struct.pack(REGISTER_V4_PACK, b"c", standalone, reg.size, len(reg_name), in_circuit)
+            )
+            buf.write(reg_name)
+            REGISTER_ARRAY_PACK = "!%sq" % reg.size
+            bit_indices = []
+            for bit in reg:
+                bit_index = clbit_indices.get(bit, -1)
+                # If index already in circuit treat it as non-existent (since it indicates dual
+                # register membership and the bit is already being reconstructed
+                if bit_index in processed_clbit_indices:
+                    bit_index = -1
+                if bit_index >= 0:
+                    processed_clbit_indices.add(bit_index)
+                bit_indices.append(bit_index)
+            buf.write(struct.pack(REGISTER_ARRAY_PACK, *bit_indices))
+
+    with io.BytesIO() as reg_buf:
+        if num_registers > 0:
+            process_qregs(reg_buf, circuit.qregs)
+            process_cregs(reg_buf, circuit.cregs)
+        # process dangling registers (those not added to circuit)
+        qreg_set = set()
+        for bit in circuit.qubits:
+            if bit._register is not None and bit._register not in circuit.qregs:
+                qreg_set.add(bit._register)
```

Comment:
I'm guessing the long-term plan here is just to remove the later loops that deal with `bit._register` once `Bit.register` is entirely removed (it's a candidate for removal in 0.20 _in theory_ I think, but we need to update our own code that does that).  Do we really need to reconstruct registers that aren't in the circuit after serialisation to make these bits work?  Back-referencing a register is already deprecated, so it feels like we'd be within our rights just to make them loose bits on serialisation, since serialisation of them is a new feature.

Don't get me wrong, this is a neat solution, though.

Reply:
Well this was mostly necessary to make equality tests pass, register initialized bits don't equal independent bits. So all the tests were failing because the bits used in the inner circuits for the controlled blocks used in the control flow circuits were evaluating as unequal. This was because the bits on the inner circuits were coming from registers but the registers are not added to those inner circuits, just the bits. As long as we are using the register to determine equality of bits we'll need this if we want to do `ser_circ == deser_circ`. That being said I don't recall the full details of the plan around the back reference deprecation, if we're just removing the user facing access or also the internal reference. @kdk can probably remind me. Either way when it is time to remove those back references some thought will need to be put into how we handle the qpy code for this, because we'll have to load circuits that had registers encoded in all of these weird patterns from previous versions of qiskit.